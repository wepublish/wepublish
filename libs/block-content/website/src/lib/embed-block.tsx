import {styled} from '@mui/material'
import {Block, EmbedBlock as EmbedBlockType} from '@wepublish/website/api'
import {BuilderEmbedBlockProps} from '@wepublish/website/builder'

export function transformCssStringToObject(styleCustom: string): Record<string, unknown> {
  const styleRules = styleCustom.split(';')
  if (styleRules.length === 0) return {}
  return styleRules.reduce((previousValue: Record<string, unknown>, currentValue: string) => {
    const [key, value] = currentValue.split(':')
    if (key && value) {
      return Object.assign(previousValue, {[key.trim()]: value.trim()})
    }
    return previousValue
  }, {})
}

export const isEmbedBlock = (block: Block): block is EmbedBlockType =>
  block.__typename === 'EmbedBlock'

export const EmbedBlockWrapper = styled('div')`
  width: 100%;
`

export function EmbedBlock({
  url,
  title,
  width,
  height,
  styleCustom,
  sandbox,
  className
}: BuilderEmbedBlockProps) {
  const ratio = width && height ? width / height : 0
  const noRatio = !!styleCustom && ratio === 0
  const styleCustomCss =
    noRatio && !!styleCustom && styleCustom !== ''
      ? transformCssStringToObject(styleCustom)
      : {
          width: '100%',
          height: '100%'
        }

  return url ? (
    <EmbedBlockWrapper className={className}>
      <div
        style={{
          position: 'relative',
          paddingTop: `${noRatio && ratio === 0 ? '0' : (1 / ratio) * 100 + '%'}`,
          minHeight: '45px'
        }}>
        <iframe
          src={url}
          title={title ?? undefined}
          allowFullScreen
          sandbox={sandbox ?? undefined}
          style={{
            position: !noRatio ? 'absolute' : 'relative',
            top: 0,
            left: 0,
            ...styleCustomCss
          }}
        />
      </div>
    </EmbedBlockWrapper>
  ) : (
    <div></div>
  )
}
