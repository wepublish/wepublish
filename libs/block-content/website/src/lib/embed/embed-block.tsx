import {styled} from '@mui/material'
import {Block, EmbedBlock as EmbedBlockType} from '@wepublish/website/api'
import {BuilderEmbedBlockProps} from '@wepublish/website/builder'
import {css} from '@emotion/react'
import {useMemo} from 'react'

export const isEmbedBlock = (block: Block): block is EmbedBlockType =>
  block.__typename === 'EmbedBlock'

export const EmbedBlockWrapper = styled('div')``

export const EmbedBlockIframe = styled('iframe')<{aspectRatio: number}>`
  width: 100%;
  border: 0;
  aspect-ratio: ${({aspectRatio}) => aspectRatio};
`

export function EmbedBlock({
  url,
  title,
  width,
  height,
  styleCustom,
  sandbox,
  className
}: BuilderEmbedBlockProps) {
  const ratio = width && height ? width / height : 1

  const styleCustomCss = useMemo(
    () => css`
      ${styleCustom}
    `,
    [styleCustom]
  )

  return url ? (
    <EmbedBlockWrapper className={className}>
      <EmbedBlockIframe
        aspectRatio={ratio}
        css={styleCustomCss}
        src={url}
        title={title ?? undefined}
        allowFullScreen
        sandbox={sandbox ?? undefined}
      />
    </EmbedBlockWrapper>
  ) : (
    <div></div>
  )
}
