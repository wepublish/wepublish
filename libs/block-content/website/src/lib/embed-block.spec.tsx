import {render} from '@testing-library/react'
import * as stories from './embed-block.stories'
import {composeStories} from '@storybook/react'

import fetch from 'jest-fetch-mock'
jest.setMock('node-fetch', fetch)

const storiesCmp = composeStories(stories)

describe('Embed Block', () => {
  Object.entries(storiesCmp).forEach(([story, Component]) => {
    it(`should render ${story}`, () => {
      const {asFragment} = render(<Component />)
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
