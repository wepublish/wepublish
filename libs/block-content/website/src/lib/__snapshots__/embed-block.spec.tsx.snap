// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Embed Block should render Default 1`] = `
<DocumentFragment>
  <div
    class="css-8atqhb e15athp30"
  >
    <div
      style="position: relative; padding-top: 56.07142857142857%; min-height: 45px;"
    >
      <iframe
        allowfullscreen=""
        sandbox=""
        src="https://www.example.com"
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
        title="Title"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Embed Block should render WithClassName 1`] = `
<DocumentFragment>
  <div
    class="extra-classname css-8atqhb e15athp30"
  >
    <div
      style="position: relative; padding-top: 56.07142857142857%; min-height: 45px;"
    >
      <iframe
        allowfullscreen=""
        sandbox=""
        src="https://www.example.com"
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
        title="Title"
      />
    </div>
  </div>
</DocumentFragment>
`;
